sfge/src/main/java/com/salesforce/graph/ops/expander/ApexPathCollector.java

```diff
@@ -7,40 +7,77 @@
 import java.util.Map;
 
 /**
- * Captures only the ApexPath of interest, allows the rest of the ApexPathExpander objects to be
- * garbage collected.
+ * Optimization to reduce the number of objects that are kept around while expanding paths. When an
+ * {@link ApexPathExpander} is accepted, its associated {@link ApexPath} is captured. When it is
+ * rejected, the {@link PathExpansionException} that caused the rejection is captured. Either way,
+ * the rest of the {@link ApexPathExpander} can be discarded and can be garbage-collected, thus
+ * relieving memory pressure.
  */
-final class ApexPathCollector extends ResultCollector {
+public final class ApexPathCollector {
     /**
      * Keep a map of the Expander Id to the ApexPath. ApexPath prevents equality comparisons to
      * avoid bugs
      */
     private final Map<Long, ApexPath> expanderIdToPath;
+    /**
+     * Mapping from the IDs of expanders whose paths were rejected to the reasons for those
+     * rejections.
+     */
+    private final Map<Long, PathExpansionException> expanderIdToRejectionReason;
 
     ApexPathCollector() {
         this.expanderIdToPath = new HashMap<>();
+        this.expanderIdToRejectionReason = new HashMap<>();
     }
 
-    List<ApexPath> getResults() {
+    public List<ApexPath> getAcceptedResults() {
         return new ArrayList<>(expanderIdToPath.values());
     }
 
-    @Override
-    boolean remove(ApexPathExpander pathExpander) {
-        Long key = pathExpander.getId();
-        return expanderIdToPath.remove(key) != null;
+    public List<PathExpansionException> getRejectionReasons() {
+        return new ArrayList<>(expanderIdToRejectionReason.values());
     }
 
-    @Override
-    void _collect(ApexPathExpander pathExpander) {
+    void collectAccepted(ApexPathExpander pathExpander) {
+        // Copy over information to each path which the expander has been accumulating.
+        // TODO: This is sharing information between the original path and initialization paths.
+        //       We're pretty sure that's fine, but we should keep an eye on it.
+        pathExpander.getTopMostPath().setApexPathMetaInfo(pathExpander.getApexPathVertexMetaInfo());
         Long key = pathExpander.getId();
         ApexPath path = pathExpander.getTopMostPath();
         expanderIdToPath.put(key, path);
         pathExpander.finished();
     }
 
-    @Override
-    int size() {
+    void collectRejected(ApexPathExpander pathExpander, PathExpansionException reason) {
+        Long key = pathExpander.getId();
+        expanderIdToRejectionReason.put(key, reason);
+        pathExpander.finished();
```

Comment:
Previously, we were only calling `.finished()` on accepted path expanders. However, since we now keep the rejection exceptions, some of which preserve the expander, it seemed prudent to call the `.finished()` method on rejected expanders too.
@rmohan20 , can you think of any potential problems that would be caused by doing this?

Reply:
Thanks for calling this out. I don't think we should call `.finished()` here. There's a good chance that more path expander candidates are waiting to be processed from the same starting point and we still need ApexPathExpander's context to handle them correctly.

Reply:
Or maybe that doesn't matter. In any case, if the ApexPathExpander isn't referenced anymore, it should get garbage collected anyway. `.finished()` is taking extra efforts to clean up. Previously, rejected paths weren't calling `finished(),` and we might as well leave it the same way.

Reply:
My concern is that `PathCollapsedException` instances have their `ApexPathExpander` as a property. So if those get collected, then their expanders won't get garbage-collected.
Proposal: That class's `getApexPathExpander()` method is never actually called. So we could remove the getter and the property on that one, and then it should be fine to remove the `finished()` call here.
Thoughts?

Reply:
@jfeingold35 I see what you mean.

Would we need to reference ApexPathExpander from the rule? Maybe not needed for NPE but if there's another rule, this could be helpful information.

Seems safer to execute performance tests before deciding where we can optimize. I can execute heap profiling while QA'g this to see if there is a significant increase, which calls for optimization to add `.finished()` or removal of reference to ApexPathExpander from exception.

Reply:
Fair enough. I'll remove the call for now, and we can add it back if need be.